File: spring-mvc-plain-java-example/src/main/java/de/cfranzen/examples/webflux/SearchByGroupController.java
package de.cfranzen.examples.webflux;

import de.cfranzen.examples.webflux.groups.CustomerGroupClient;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
class SearchByGroupController {

    private final CustomerGroupClient client;

    private final CustomerRepository repository;

    SearchByGroupController(final CustomerGroupClient client, final CustomerRepository repository) {
        this.client = client;
        this.repository = repository;
    }

    @GetMapping("/search/customerGroup/{customerGroup}")
    List<Customer> findByCustomerGroup(@PathVariable final String customerGroup) {
        final var group = client.getCustomerGroup(customerGroup);
        return group.entries()
                .stream()
                .flatMap(entry -> repository.findByFirstNameAndLastName(entry.firstName(), entry.lastName()))
                .toList();
    }
}

File: spring-webflux-kotlin-example/src/main/kotlin/de/cfranzen/examples/webflux/kotlin/SearchByGroupController.kt
package de.cfranzen.examples.webflux.kotlin

import de.cfranzen.examples.webflux.kotlin.groups.CustomerGroupClient
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.merge
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RestController

@RestController
internal class SearchByGroupController(
    private val client: CustomerGroupClient,
    private val repository: CustomerRepository
) {

    @GetMapping("/search/customerGroup/{customerGroup}")
    fun findByCustomerGroup(@PathVariable customerGroup: String) = flow {
        val group = client.getCustomerGroup(customerGroup)
        emitAll(group
            .entries
            .map { repository.findByFirstNameAndLastName(it.firstName, it.lastName) }
            .merge())
    }
}
